### DISCUSSION TITLE
[Curso] Fundamentos do Aspnet 6 [Aula] Testando Autenticação e Autorização

### DISCUSSION THREAD

**Discussion by paulamora-is** [balta-io/forum#364] [Tira Dúvidas] [answered]
Olá, Balta, tudo bem? Segui os mesmos passos, mas na hora de testar no postman, recebo Unauthorized mesmo contendo as roles de autorização. O que pode ser?

 **Controller**
       
       
        [Authorize(Roles = "user")]
        [HttpGet("v1/user")]
        public IActionResult GetUser() => Ok(User.Identity.Name);

        [Authorize(Roles = "author")]
        [HttpGet("v1/author")]
        public IActionResult GetAuthor() => Ok(User.Identity.Name);

        [Authorize(Roles = "admin")]
        [HttpGet("v1/admin")]
        public IActionResult GetAdmin() => Ok(User.Identity.Name);

**Program**

```
var key = Encoding.ASCII.GetBytes(Configuration.JwtKey);
      builder.Services.AddAuthentication(options =>
      {
          options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
          options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
      }).AddJwtBearer(options =>
      {
          options.TokenValidationParameters = new TokenValidationParameters
          {
              ValidateIssuerSigningKey = true,
              IssuerSigningKey = new SymmetricSecurityKey(key),
              ValidateIssuer = false,
              ValidateActor = false
          };
      });
```

```
app.UseAuthentication();
app.UseAuthorization();
```

**TokenService**

```
public class TokenService
    {
        public string GenerateToken(User user)
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            var key = Encoding.ASCII.GetBytes(Configuration.JwtKey);
            var tokenDescriptor = new SecurityTokenDescriptor
            {
                Subject = new ClaimsIdentity(new Claim[]
                {
                    new Claim(ClaimTypes.Name, "paulamorais"),
                    new Claim(ClaimTypes.Role, "admin"),
                    new Claim(ClaimTypes.Role, "user")
                }),

                Expires = DateTime.UtcNow.AddHours(8),
                SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
            };

            var token = tokenHandler.CreateToken(tokenDescriptor);
            return tokenHandler.WriteToken(token);
        }
    }

```

![image](https://github.com/balta-io/forum/assets/55562741/7352b064-2e68-4a3a-bf96-c925aee5bd1f)

Meus pacotes: 

![image](https://github.com/balta-io/forum/assets/55562741/30523b1c-3799-4ce7-ac24-cf39a8a17e4a)

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @paulamora-is, como vai?
Poderia rodar os fontes do curso para comparar e ver onde está diferente? Eu testei bem este exemplo e acredito que esteja funcionando...

Deve ser algum detalhe, talvez no nome do Role!
Você chegou a debugar pra ver se o User.Identity vem preenchido? Se é um problema nele ou nos roles?

**Comment by sinvalfelisberto** [comment]
Balta, beleza?
Tive aqui o mesmo problema!

Tinha errado aqui, na program!

app.UseAuthentication();
app.UseAuthorization();

Em vez de setar o UseAuthentication() e depois do UseAuthorization(), tinha colocado o UseAuthentication() duas vezes... não dá erro, permite o AllowAnonnymous, mas dá o mesmo retorno (401) que o colega estava enfrentando.

Lutei um pouco pra conseguir encontrar, mas consegui!
Valeu demais pelo curso! Estou aprendendo um monte de coisa massa!
